Allow selecting a default account to use with an instance

byteduck opened the pull request:
This commit adds the functionality to select a default account to use when launching an instance (#341). If no account is supplied via the accountToUse parameter of `Application::launch`, then the instance will launched logged into the default account for the instance (or the global default account if none is set for the instance).

The setting is found in the Miscellaneous tab of the instance settings and looks like this:
![image](https://user-images.githubusercontent.com/10217842/209726879-364d5808-7a73-4018-b942-15d2524cfb02.png)
![image](https://user-images.githubusercontent.com/10217842/209726906-6bba24ba-c6b1-4bce-83c6-953581b7af9f.png)

I couldn't manage to get a screenshot with the dropdown open, but it contains an option for every account configured in Prism. To revert the setting, one can simply uncheck the "Set a default account to use with this instance" box. If the account configured is removed, then the setting is reverted to showing "No default account".

byteduck committed:
Allow selecting a default account to use with an instance

Signed-off-by: Aaron <10217842+byteduck@users.noreply.github.com>

byteduck committed:
Fix conflicting layout name in InstanceSettingsPage

Signed-off-by: Aaron <10217842+byteduck@users.noreply.github.com>

flowln commented on launcher/ui/pages/instance/InstanceSettingsPage.cpp:
```diff
@@ -437,6 +454,70 @@ void InstanceSettingsPage::on_javaTestBtn_clicked()
+
+    // Profile's associated Mojang username
+    if (sAction->data().type() != QVariant::Type::Int)
+        return;
```
```suggestion
    Q_ASSERT(sAction->data().type() == QVariant::Type::Int);
```
seems to me that this should never not be an integer

flowln commented on launcher/ui/pages/instance/InstanceSettingsPage.cpp:
```diff
@@ -437,6 +454,70 @@ void InstanceSettingsPage::on_javaTestBtn_clicked()
+        ui->instanceAccountSelector->setText(account->profileName());
+        ui->instanceAccountSelector->setIcon(account->getFace());
+    } else {
+        ui->instanceAccountSelector->setText(tr("No default account"));
+        ui->instanceAccountSelector->setIcon(APPLICATION->getThemedIcon("noaccount"));
```
I'm not sure what this means. Even with the instance-specific override, leaving this selected doesn't prompt the user to select an account at launch (as one would expect from a 'No default account' option). It simply uses the more generic launcher-wide default account. Maybe this should be rephrased?

byteduck replied:
As @TheKodeToad pointed out, the text for the checkbox was a little verbose. I reworded the checkbox to "Override default account", and when unchecked, the greyed-out account dropdown will mirror the global dropdown in the main launcher window (ie, it will show the currently selected default account or show "No default account" if none is selected)

flowln commented on launcher/ui/pages/instance/InstanceSettingsPage.cpp:
```diff
@@ -437,6 +454,70 @@ void InstanceSettingsPage::on_javaTestBtn_clicked()
+    m_settings->set("InstanceAccountId", account->profileId());
+
+    ui->instanceAccountSelector->setText(account->profileName());
+    ui->instanceAccountSelector->setIcon(account->getFace());
```
```suggestion
    if (auto face = account->getFace(); !face.isNull()) {
        ui->instanceAccountSelector->setIcon(face);
    } else {
        ui->instanceAccountSelector->setIcon(APPLICATION->getThemedIcon("noaccount"));
    }
```

flowln commented on launcher/ui/pages/instance/InstanceSettingsPage.cpp:
```diff
@@ -437,6 +454,70 @@ void InstanceSettingsPage::on_javaTestBtn_clicked()
+    bool valid = false;
+    int index = data.toInt(&valid);
+    if(!valid) {
+        index = -1;
```
you're already testing if the QVariant is an integer up there, so this will always be valid, right? Also, wouldn't the `at()` call fail with a negative index?

byteduck replied:
You're right - I copied this code from the main account switcher and didn't notice that. Removed the check. The account selected from the dropdown should always have a valid index anyway.

flowln commented on launcher/ui/pages/instance/InstanceSettingsPage.cpp:
```diff
@@ -437,6 +454,70 @@ void InstanceSettingsPage::on_javaTestBtn_clicked()
+        }
+
+        accountMenu->addAction(action);
+        connect(action, SIGNAL(triggered(bool)), SLOT(changeInstanceAccount()));
```
```suggestion
        connect(action, SIGNAL(triggered(bool)), this, SLOT(changeInstanceAccount()));
```

flowln commented on launcher/ui/pages/instance/InstanceSettingsPage.cpp:
```diff
@@ -75,6 +80,7 @@ bool InstanceSettingsPage::shouldDisplay() const
 InstanceSettingsPage::~InstanceSettingsPage()
 {
     delete ui;
+    delete accountMenu;
```
```suggestion
```
since you've created the QMenu with a parent, it will be deleted with the window anyway, you don't have to do it manually.

flowln reviewed (commented):
seems to be working mostly fine, thanks! most the suggested changes are only about some code cleanups :)

we have a `.clang-format` thingy in the launcher's root directory, please use that on the code you've added, the formatting is somewhat inconsistent :p

TheKodeToad commented:
imo the option name ~~name~~ is slightly longer than it should be.

byteduck commented:
> imo the option name ~~name~~ is slightly longer than it should be.

As in, the text seen in the GUI or the key used for the settings store?

byteduck commented:
> we have a `.clang-format` thingy in the launcher's root directory, please use that on the code you've added, the formatting is somewhat inconsistent :p

Ah yes, my apologies - I was trying to follow the style of the code I saw and didn't notice the .clang-format. Will do!

byteduck committed:
Replace unecessary type check with assertion in InstanceSettingsPage

Co-authored-by: flow <flowlnlnln@gmail.com>
Signed-off-by: Aaron Sonin <10217842+byteduck@users.noreply.github.com>
Signed-off-by: Aaron <10217842+byteduck@users.noreply.github.com>

byteduck committed:
Add null check for face in instance account settings selector

Co-authored-by: flow <flowlnlnln@gmail.com>
Signed-off-by: Aaron Sonin <10217842+byteduck@users.noreply.github.com>
Signed-off-by: Aaron <10217842+byteduck@users.noreply.github.com>

TheKodeToad commented:
> > imo the option name ~name~ is slightly longer than it should be.
> 
> As in, the text seen in the GUI or the key used for the settings store?

In the GUI. "for this instance" could be seen as redundant.

byteduck committed:
Properly connect signal in instance settings for account selector

Co-authored-by: flow <flowlnlnln@gmail.com>
Signed-off-by: Aaron Sonin <10217842+byteduck@users.noreply.github.com>
Signed-off-by: Aaron <10217842+byteduck@users.noreply.github.com>

byteduck committed:
Remove unecessary delete in InstanceSettingsPage destructor

Co-authored-by: flow <flowlnlnln@gmail.com>
Signed-off-by: Aaron Sonin <10217842+byteduck@users.noreply.github.com>
Signed-off-by: Aaron <10217842+byteduck@users.noreply.github.com>

byteduck committed:
Reword instance-specific account settings, apply clang-format

Signed-off-by: Aaron <10217842+byteduck@users.noreply.github.com>

byteduck commented:
> > > imo the option name ~name~ is slightly longer than it should be.
> > 
> > 
> > As in, the text seen in the GUI or the key used for the settings store?
> 
> In the GUI. For this instance could be seen as redundant.

I've reworded it to "Override default account". When unchecked, the instance-specific account selector will show the current global setting for the default account.

byteduck committed:
Merge branch 'PrismLauncher:develop' into instance-accounts

byteduck commented:
Hmm, not sure why the `ResourceFolderModel` test failed on `Windows-MSVC-Legacy` only. Nothing I did should affect that, right? I did notice that `Windows-MSVC-Legacy` failed when I first opened this PR, and when run a second time after signing my commit, it succeeded.

flowln commented:
> Hmm, not sure why the `ResourceFolderModel` test failed on `Windows-MSVC-Legacy` only. Nothing I did should affect that, right? I did notice that `Windows-MSVC-Legacy` failed when I first opened this PR, and when run a second time after signing my commit, it succeeded.

seems to be some issue on develop, i've seen another PR with the same error earlier today, dw about that

flowln commented on launcher/ui/pages/instance/InstanceSettingsPage.cpp:
```diff
@@ -437,6 +452,67 @@ void InstanceSettingsPage::on_javaTestBtn_clicked()
+
+    if (defaultAccount) {
+        ui->instanceAccountSelector->setText(defaultAccount->profileName());
+        ui->instanceAccountSelector->setIcon(defaultAccount->getFace());
```
there is a small cornercase here regarding Offline accounts: Since they don't have an Icon, the returned `QPixmap` for the `getFace()` call is null. So, this doesn't set the icon the first time the page is opened and one of those is set. You'd probably need to do something like

```cpp
if (auto face = defaultAccount->getFace(); !face.isNull())
    ui->instanceAccountSelector->setIcon(face);
else
    ui->instanceAccountSelector->setIcon(APPLICATION->getThemedIcon("noaccount"));
```

for it to work as intended.

Considering this very same piece of code is repeated in other places here to, it may make sense to put this logic in a separate function :>

byteduck replied:
Good thinking. Do you think this logic should be put in `MinecraftAccount`, named something like `getFaceIcon`?

flowln replied:
It can just be a simple function in `InstanceSettingsPage.cpp` i think, no need to get fancy, we don't use this logic anywhere else (i think?)

byteduck replied:
Fixed! I also extracted the logic into a method called `getFaceForAccount` and used it where applicable :)

flowln reviewed (approved):
sorry, i kinda forgot i didn't already review this :sweat_smile: 

anyway, the code looks mostly good, there's just a little cornercase and maybe moving some logic to a helper function, otherwise it's all good! thanks!

byteduck commented:
> sorry, i kinda forgot i didn't already review this :sweat_smile: 

No worries! I'll get the corner case fixed soon :)

byteduck committed:
Fix instance account selector face for offline accounts

byteduck committed:
Merge branch 'PrismLauncher:develop' into instance-accounts

byteduck committed:
DCO Remediation Commit for Aaron <10217842+byteduck@users.noreply.github.com>

I, Aaron <10217842+byteduck@users.noreply.github.com>, hereby add my Signed-off-by to this commit: 160dd09fc2788fea17c8e9e332c2877586640971

Signed-off-by: Aaron <10217842+byteduck@users.noreply.github.com>

byteduck commented:
Seems like some Ubuntu package mirrors are down which caused the checks to fail... I'm encountering this on my own projects too :grimacing:

DioEgizio commented:
> Seems like some Ubuntu package mirrors are down which caused the checks to fail... I'm encountering this on my own projects too 😬

https://github.com/actions/runner-images/issues/6894

flowln reviewed (approved):
ideally we'd have some place in the UI to indicate that the account for a given instance is overridden. However, that'd probably mean messing with the account selector thingy, so it's probably a bit troublesome for now...

anyway, this looks good to me for the time being, so thank you! :D

Scrumplex reviewed (approved):
LGTM, thanks!
